### ISSUE TITLE
수요예측 완성품

### ISSUE THREAD

**Issue by 0redbluefin0** [0redbluefin0/suyo#2]
import numpy as np
import pandas as pd
from io import StringIO
import xgboost as xgb
import matplotlib.pyplot as plt

# =========================
# 1. 데이터 로드
# =========================
csv_data = """year,is_seollal_flag,chuseok_flag,holiday_length,concurrent_connections,seat_count
2022,1,0,6,290000,530398
2022,0,1,5,830000,978771
2023,1,0,5,160000,977891
2023,0,1,7,620000,1251869
2024,1,0,5,420000,800153
2024,0,1,6,500000,985627
2025,1,0,10,480000,1625660
2025,0,1,11,1850000,1722983
2026,1,0,5,,1000000
"""

df = pd.read_csv(StringIO(csv_data))

# 숫자형 변환
num_cols = [
    'year',
    'is_seollal_flag',
    'chuseok_flag',
    'holiday_length',
    'concurrent_connections',
    'seat_count',
]
for col in num_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# 정렬
df = df.sort_values(['is_seollal_flag','chuseok_flag','year']).reset_index(drop=True)

# =========================
# 2. 이전 연도 피처 생성
# =========================
df['prev_concurrent_connections'] = np.nan
df['prev_seat_count'] = np.nan

for sel in [0, 1]:
    for chu in [0, 1]:
        idx = df[(df['is_seollal_flag'] == sel) & (df['chuseok_flag'] == chu)].index
        if len(idx) == 0:
            continue
        sub = df.loc[idx].sort_values('year')
        df.loc[idx, 'prev_concurrent_connections'] = sub['concurrent_connections'].shift(1).values
        df.loc[idx, 'prev_seat_count'] = sub['seat_count'].shift(1).values

df['concurrent_growth_vs_prev'] = (
    df['concurrent_connections'] - df['prev_concurrent_connections']
) / df['prev_concurrent_connections']
df['seat_growth_vs_prev'] = (
    df['seat_count'] - df['prev_seat_count']
) / df['prev_seat_count']

# =========================
# 3. Capacity 수식 정의
# =========================
BASE_LENGTH = 5.0  # 기준 연휴 길이 (5일)

# 2022년 추석 (is_seollal_flag=0, chuseok_flag=1)을 기준으로 base_factor 추정
row_2022_chuseok = df[
    (df['year'] == 2022) &
    (df['is_seollal_flag'] == 0) &
    (df['chuseok_flag'] == 1)
].iloc[0]

user_2022 = row_2022_chuseok['concurrent_connections']
seat_2022 = row_2022_chuseok['seat_count']
length_2022 = row_2022_chuseok['holiday_length']

# 2022 추석 효율을 0.90 정도로 가정
assumed_eff_2022 = 0.90
capacity_2022 = user_2022 / assumed_eff_2022

base_factor = capacity_2022 / (seat_2022 * (length_2022 / BASE_LENGTH))

print(f"[INFO] 추정 base_factor: {base_factor:.4f}")

# 전체 행에 capacity 계산
df['capacity'] = base_factor * df['seat_count'] * (df['holiday_length'] / BASE_LENGTH)

# =========================
# 4. Efficiency 계산 및 모델 학습
# =========================
# 실측 concurrent_connections가 있는 행만 사용 (2026 설은 NaN이라 자동 제외)
df_eff = df[~df['concurrent_connections'].isna()].copy()

df_eff['efficiency'] = df_eff['concurrent_connections'] / df_eff['capacity']
# 효율 이상치 방지용 클리핑
df_eff['efficiency'] = df_eff['efficiency'].clip(lower=0.3, upper=1.2)

feature_cols = [
    'year',
    'is_seollal_flag',
    'chuseok_flag',
    'holiday_length',
    'seat_count',
    'prev_concurrent_connections',
    'prev_seat_count',
    'concurrent_growth_vs_prev',
    'seat_growth_vs_prev',
]

X_eff_train = df_eff[feature_cols].values
y_eff_train = df_eff['efficiency'].values

eff_model = xgb.XGBRegressor(
    objective='reg:squarederror',
    n_estimators=300,
    learning_rate=0.1,
    max_depth=3,
    subsample=0.9,
    colsample_bytree=0.9,
    random_state=42,
)
eff_model.fit(X_eff_train, y_eff_train)

# =========================
# 5. 2026년 설날 concurrent 예측
# =========================
row_2026_seollal = df[
    (df['year'] == 2026) &
    (df['is_seollal_flag'] == 1) &
    (df['chuseok_flag'] == 0)
].iloc[0]

X_2026 = row_2026_seollal[feature_cols].values.reshape(1, -1)
eff_2026_pred = float(eff_model.predict(X_2026)[0])
eff_2026_pred = float(np.clip(eff_2026_pred, 0.5, 1.1))  # 현실적인 효율 범위로 제한

capacity_2026 = row_2026_seollal['capacity']
concurrent_2026_pred = capacity_2026 * eff_2026_pred

print("\n=== 2026년 설날 concurrent_connections 예측 (Capacity × Efficiency) ===")
print(f"2026 설 capacity (이론 상한) : {capacity_2026:,.0f}")
print(f"2026 설 efficiency 예측값   : {eff_2026_pred:.2f}")
print(f"2026 설 concurrent 예측값   : {concurrent_2026_pred:,.0f}")

# df에 2026 설 예측 concurrent 반영
df.loc[
    (df['year'] == 2026) &
    (df['is_seollal_flag'] == 1) &
    (df['chuseok_flag'] == 0),
    'concurrent_connections'
] = concurrent_2026_pred

# efficiency 다시 계산 (예측치 포함)
df['efficiency'] = df['concurrent_connections'] / df['capacity']

# =========================
# 6. 추석만 별도 요약 테이블 (2022~2025, 2026은 데이터 없음)
# =========================
df_chuseok = df[df['chuseok_flag'] == 1].copy().sort_values('year')
print("\n=== 추석 요약 ===")
print(df_chuseok[['year', 'holiday_length', 'concurrent_connections',
                  'seat_count', 'capacity', 'efficiency']])

# 설날만 요약 (2022~2026)
df_seollal = df[(df['is_seollal_flag'] == 1) & (df['chuseok_flag'] == 0)].copy().sort_values('year')
print("\n=== 설날 요약 ===")
print(df_seollal[['year', 'holiday_length', 'concurrent_connections',
                  'seat_count', 'capacity', 'efficiency']])

# =========================
# 7. 그래프 1: 추석 concurrent vs seat_count 콤보 차트
# =========================
if len(df_chuseok) > 0:
    years_c = df_chuseok['year'].astype(int).tolist()
    concurrent_c = df_chuseok['concurrent_connections'].tolist()
    seats_c = df_chuseok['seat_count'].tolist()

    fig, ax1 = plt.subplots(figsize=(8, 5))
    ax2 = ax1.twinx()

    # 막대: concurrent_connections
    bars = ax1.bar(years_c, concurrent_c, color='#1f77b4', alpha=0.8, label='Concurrent Connections')

    # 선: seat_count
    ax2.plot(years_c, seats_c, color='#ff7f0e', marker='o', linewidth=2, label='Seat Count')

    # 2025 강조 (실측이지만 최신값이라 강조)
    for i, y in enumerate(years_c):
        if y == 2025:
            bars[i].set_color('#4daf4a')  # 초록색 강조

    ax1.set_xlabel('Year')
    ax1.set_ylabel('Concurrent Connections')
    ax2.set_ylabel('Seat Count')
    ax1.set_title('Chuseok: Concurrent Connections vs Seat Count (2022–2025)')

    # 범례
    lines_1, labels_1 = ax1.get_legend_handles_labels()
    lines_2, labels_2 = ax2.get_legend_handles_labels()
    ax1.legend(lines_1 + lines_2, labels_1 + labels_2, loc='upper left')

    plt.tight_layout()
    plt.show()
    # plt.savefig('chuseok_concurrent_vs_seats.png', dpi=150)

# =========================
# 8. 그래프 2: Capacity vs Actual Concurrent (추석)
# =========================
if len(df_chuseok) > 0:
    years_c = df_chuseok['year'].astype(int).tolist()
    capacity_vals_c = df_chuseok['capacity'].tolist()
    concurrent_vals_c = df_chuseok['concurrent_connections'].tolist()

    x = np.arange(len(years_c))
    width = 0.35

    fig, ax = plt.subplots(figsize=(8, 5))

    bar1 = ax.bar(x - width/2, capacity_vals_c, width,
                  label='Capacity (Theoretical Max)', color='#bbbbbb')
    bar2 = ax.bar(x + width/2, concurrent_vals_c, width,
                  label='Actual Concurrent', color='#1f77b4')

    # 2025 막대 강조
    for i, y in enumerate(years_c):
        if y == 2025:
            bar2[i].set_color('#4daf4a')  # 초록색

    ax.set_xticks(x)
    ax.set_xticklabels(years_c)
    ax.set_ylabel('Number of Connections')
    ax.set_title('Chuseok: Capacity vs Actual Concurrent')
    ax.legend()

    plt.tight_layout()
    plt.show()
    # plt.savefig('chuseok_capacity_vs_concurrent.png', dpi=150)

# =========================
# 9. 그래프 3: Efficiency 추이 (추석)
# =========================
if len(df_chuseok) > 0:
    years_c = df_chuseok['year'].astype(int).tolist()
    eff_vals_c = df_chuseok['efficiency'].tolist()

    fig, ax = plt.subplots(figsize=(8, 5))
    ax.plot(years_c, eff_vals_c, marker='o', color='#ff7f0e', linewidth=2)

    for y, e in zip(years_c, eff_vals_c):
        label = f"{e:.2f}"
        ax.text(y, e + 0.02, label, ha='center', va='bottom', fontsize=9)

    ax.set_xlabel('Year')
    ax.set_ylabel('Efficiency (Actual/Capacity)')
    ax.set_title('Chuseok: Efficiency Trend')
    ax.set_ylim(0.3, 1.2)

    plt.grid(axis='y', linestyle='--', alpha=0.5)
    plt.tight_layout()
    plt.show()
    # plt.savefig('chuseok_efficiency_trend.png', dpi=150)
